### PR TITLE
refactor integration tests

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why is this change needed?

The integration tests became quite hard to follow. 
There is a general pattern for all tests except the pure in-memory test:

1. Start a geth network , deploy the contracts, and return the addresses
2. Start up obscuro components
  a) Enclaves. Which can be started inside the obscuro node, standalone process, inside docker or manually in azure
  b) The Obscuro node which can be in memory or standalone and it has to know the address of the enclave.

The stopping order should be:
- first stop the obscuro host. Because this glues everything together, and needs everything up and running until the very end.
- then stop the enclaves and the ethereum network

### What changes were made as part of this PR:
This is mostly a refactoring PR. The only thing that changed functionally is the sequence of stopping resources in TearDown. 

- the geth setup and teardown logic was extracted in functions and move to a file
- starting enclaves was extracted and the rpc addresses of the enclaves are now being passed to the logic that starts the nodes

These changes allow for the `Network.Create` functions to be much more compact

- the `TearDown` now implements the stopping pattern

- the "ethclient.EthClient" interface was being passed around as a pointer.

### What are the key areas to look at
 - The readability of the "azureenclave", "dockerenclave",  "geth_network" and "socket" network setup files

### CODE REVIEW

**Review by BedrockSquirrel — COMMENTED** [review]
Bit unclear on what the problem is that these changes address and how it addresses it. Maybe we need to avoid spawning goroutines in the tear down if order matters.

File: integration/simulation/network/dockerenclave.go [hunk 1]
```diff
@@ -210,17 +210,17 @@ func (n *basicNetworkOfNodesWithDockerEnclave) Create(params *params.SimParams,
 }
 
 func (n *basicNetworkOfNodesWithDockerEnclave) TearDown() {
-	n.gethNetwork.StopNodes()
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply integration/simulation/network/dockerenclave.go:213]: not sure what you mean.
The intent here is to stop the obscuro nodes first, and then stop the geth nodes

[the remark this answers is not in the document]
Comment (BedrockSquirrel) [reply integration/simulation/network/dockerenclave.go:213]: the for loop in this TearDown just firest off some goroutines then immediately proceeds to stop run this StopNodes command. So won't it introduce a race condition if the order is important?

File: integration/simulation/network/geth_network.go [hunk 2]
```diff
@@ -142,14 +142,14 @@ func (n *networkInMemGeth) Create(params *params.SimParams, stats *stats.Stats)
 }
 
 func (n *networkInMemGeth) TearDown() {
-	defer n.gethNetwork.StopNodes()
```
[the remark this answers is not in the document]
Comment (BedrockSquirrel) [reply integration/simulation/network/geth_network.go:145]: Ok cool, so the removing of the defers is because they seem unnecessary/confusing rather than to fix the issues, that's fair. I put them in because we were paranoid about panics causing things not to execute but seems unlikely in this method.

File: integration/simulation/network/dockerenclave.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tudor-malene) [reply integration/simulation/network/dockerenclave.go:213]: good point. I'll have another look tomorrow. Pretty tired right now
